pkg/ring/ring.go

```diff
@@ -340,24 +344,27 @@ func (r *Ring) updateRingState(ringDesc *Desc) {
 }
 
 // Get returns n (or more) instances which form the replicas for the given key.
-func (r *Ring) Get(key uint32, op Operation, bufDescs []InstanceDesc, bufHosts, bufZones []string) (ReplicationSet, error) {
+func (r *Ring) Get(key uint32, op Operation, bufDescs []InstanceDesc, bufHosts []string, bufZones map[string]int) (ReplicationSet, error) {
```

Comment:
Before the change:

```
% go test -benchmem -run=^$ -bench ^BenchmarkRing_Get$ github.com/cortexproject/cortex/pkg/ring
goos: darwin
goarch: amd64
pkg: github.com/cortexproject/cortex/pkg/ring
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkRing_Get-12    	 1389476	       831.3 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/ring	2.560s
```

After the change:

```
% go test -benchmem -run=^$ -bench ^BenchmarkRing_Get$ github.com/cortexproject/cortex/pkg/ring
goos: darwin
goarch: amd64
pkg: github.com/cortexproject/cortex/pkg/ring
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkRing_Get-12    	 1319292	       890.8 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/ring	2.628s
```

I don't think there's noticeable difference

Reply:
Probably better to benchmark `benchmarkBatch` because that's what called by Distributor and where you allocate memory for the map every time.
`BenchmarkRing_Get` reuses the same map so cannot tell mem alloc.

Reply:
Hmm all the `BenchmarkBatch` seem to be broken, even in the main branch. Let me see if I can fix it

Reply:
kk fixed it, here's the result before this PR

```
% go test -benchmem -run=^$ -bench ^BenchmarkBatch github.com/cortexproject/cortex/pkg/ring
goos: darwin
goarch: amd64
pkg: github.com/cortexproject/cortex/pkg/ring
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkBatch10x100-12      	   12736	     92322 ns/op	   20098 B/op	      51 allocs/op
BenchmarkBatch100x100-12     	    8013	    151404 ns/op	   56472 B/op	     431 allocs/op
BenchmarkBatch100x1000-12    	    1176	   1026580 ns/op	  191259 B/op	     420 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/ring	5.703s
```

and after this PR

```
% go test -benchmem -run=^$ -bench ^BenchmarkBatch github.com/cortexproject/cortex/pkg/ring
goos: darwin
goarch: amd64
pkg: github.com/cortexproject/cortex/pkg/ring
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkBatch10x100-12      	   12307	     94764 ns/op	   20077 B/op	      51 allocs/op
BenchmarkBatch100x100-12     	    8028	    148776 ns/op	   56465 B/op	     431 allocs/op
BenchmarkBatch100x1000-12    	    1173	   1062618 ns/op	  191209 B/op	     420 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/ring	5.782s
```

Again, no noticeable difference

Reply:
Nice thanks!